utils/core/src/serde/mod.rs

```diff
@@ -0,0 +1,146 @@
+use super::{flatten_slice_elements, DeserializationError, Vec};
+
+mod byte_reader;
+pub use byte_reader::{ByteReader, SliceReader};
+
+mod byte_writer;
+pub use byte_writer::ByteWriter;
+
+// SERIALIZABLE TRAIT
+// ================================================================================================
+
+/// Defines how to serialize `Self` into bytes.
+pub trait Serializable: Sized {
+    // REQUIRED METHODS
+    // --------------------------------------------------------------------------------------------
+    /// Serializes `self` into bytes and writes these bytes into the `target`.
+    fn write_into<W: ByteWriter>(&self, target: &mut W);
+
+    // PROVIDED METHODS
+    // --------------------------------------------------------------------------------------------
+
+    /// Serializes `self` into a vector of bytes.
+    fn to_bytes(&self) -> Vec<u8> {
+        let mut result = Vec::with_capacity(self.get_size_hint());
+        self.write_into(&mut result);
+        result
+    }
+
+    /// Serializes all elements of the `source` and writes these bytes into the `target`.
+    ///
+    /// This method does not write any metadata (e.g. number of serialized elements) into the
+    /// `target`.
+    fn write_batch_into<W: ByteWriter>(source: &[Self], target: &mut W) {
+        for item in source {
```

Comment:
We probably want to serialize the length of the collection before the data as it could be impossible to reconstruct the collection from a stream of bytes if we don't do that.

Example
```rust
(source.len() as u64).write_into(target);
```

It is important we cast to a specific type as `usize` might differ depending on architecture, and we want to avoid architecture-dependant serialization.

Reply:
Here, we don't want to serialize the length as we don't know what are potential maximum bounds on the length or if the length can be determined in some other way.

So, the user of this method is responsible for either serialization of length or convening length info to the deserializer in some other way.